File: FinalAssessment_19459138/Question4/FA_StackTest.java
/**
 * DSA Final Assessment Question 4 - FA_StackTest.java                             4
 *
 * Name : Connor Kuljis	
 * ID   : 19459138
 *
 * tests 2,4 have been removed as inbuilt stack has no limit in the constructor, it is implemented as a list
 * 	a limit can be set by creating own helper function in class if neccessary
 * 	referenced for methods : https://docs.oracle.com/javase/7/docs/api/java/util/Stack.html
 **/
import java.util.*;


public class FA_StackTest
{
    public static void main(String[] args) 
    {

        // Test Variables
        int stackSize = 50;
        int testsPassed = 0;
        int numTests = 0; 
        // FA_Stack testStack = null;
	Deque<Integer> testStack;
        Object stackElement;

        System.out.println("\n**** Question $: Testing Built-in Stacks ****\n");

	testStack = new ArrayDeque<Integer>();
        try
        {
            numTests++;
            System.out.print( "Test 1 - stack created             : " );

            // if ( testStack.getCount() != 0 )
             if ( testStack.size() != 0 )
            {
                throw new PracExamException( "Error: getCount should be 0" );
            }

            if ( !testStack.isEmpty() )
            {
                throw new PracExamException( "Error: isEmpty should be true" );
            }

	    /*
	     * there is no isFull function as it inherits as a abstract list eg (no maximum size of list)
            if ( testStack.isFull() )
            {
                throw new PracExamException( "Error: isFull should be false" );
            }
	    */

            testsPassed++;
            System.out.println( "Passed" );
        }
        catch(PracExamException e)
        {
            System.out.println( "Failed" + e.getMessage() );
        } 

	/* CANNOT SET MAX SIZE OF STACK - WITHOUT USE OF HELPER FUNCION
        try
        {
            numTests++;

            System.out.print( "Test 2 - stack created set size    : " );
            testStack = new FA_Stack(stackSize);

            if ( testStack.getCount() != 0 )
            {
                throw new PracExamException( "Error: getCount should be 0" );
            }

            if ( !testStack.isEmpty() )
            {
                throw new PracExamException( "Error: isEmpty should be true" );
            }

            if ( testStack.isFull() )
            {
                throw new PracExamException( "Error: isFull should be false" );
            }

            testsPassed++;
            System.out.println( "Passed" );
        }
        catch(PracExamException e)
        {
            System.out.println( "Failed" + e.getMessage() );
        } 
	*/

        try 
        {
            numTests++;

            System.out.print( "Test 3 - pushing values     (10,20,30)       : " );

            testStack.push( 10 );
            testStack.push( 20 );
            testStack.push( 30 );

            // if ( !testStack.top().equals(30) )
            if ( !testStack.peek().equals(30) )
            {
                throw new PracExamException( 
                        "top() returned incorrect element" );
            }
            testsPassed++;
            System.out.println( "Passed" );
		}
        catch ( PracExamException e ) {
            System.out.println( "Test " + numTests + " Failed: "+ e.getMessage() );
        }

	/* there is no limit on the built in stack
	 *
        try 
        {
	    testStack = new ArrayDeque<Integer>();
            numTests++;

            System.out.print( "Test 4 - pushing values to limit   : " );

            for ( int ii = 3; ii <= testStack.size(); ii++ )
            {
                testStack.push( ii );
            }

            testStack.push( 100 );
            System.out.println( "Test " + numTests + " Failed: push() didn't throw exception" );

        } 
        catch( PracExamException e )
        {
                testsPassed++;
                System.out.println( "Passed" );
        }
	*/

        try 
        {
            numTests++;
            System.out.print( "Test 5 - popping values            : " );

            System.out.print(testStack.pop()); // testing the pop method
            testsPassed++; // if we get to here the pop operation was successful
            System.out.println( "Passed" );
	}
        catch( Exception e ) // go here if pop() encounters exception
        {
             System.out.println( "Test " + numTests + " Failed: pop() didn't throw exception" );
        }
        
        


        System.out.println( "------ FA Stack Test Harness ------" );
        System.out.println( "Number of Tests:        " + numTests );
        System.out.println( "Number of Tests Passed: " + testsPassed );
        System.out.println( "              % Passed: " + 100*testsPassed/numTests + "%");
        System.out.println("\n**** Tests Complete ****\n");


    }

}

File: FinalAssessment_19459138/Question4/FA_StackTest.py
#
# DSA Final Assessment Question 4 - FA_StackTest.py
#
# Name : 
# ID   :
#
# 
from FA_Stack import *
from PracExamException import *

print(“\n**** Question 4: Testing Built-in Stack ****\n”)

mystack = FA_Stack()
testPassed = 0
numTests = 0

numTests += 1
print("Test",numTests,"- stack created and starts at count = 0", end="")
if mystack.getCount() == 0:
    testPassed += 1
    print("\t - PASSED")    

numTests += 1
print("Test",numTests,"- starts empty", end="")
if mystack.isEmpty():
    testPassed += 1
    print("\t\t\t\t - PASSED")    

mystack.push(10)
mystack.push(20)
mystack.push(30)

numTests += 1
print("Test",numTests,"- has three values after 3 pushes", end="")
if mystack.getCount() == 3:
    testPassed += 1
    print("\t - PASSED")    
    
numTests += 1
print("Test",numTests,"- stack is not empty", end="")
if not mystack.isEmpty():
    testPassed += 1
    print("\t\t\t - PASSED")    

pop1 = mystack.pop()
pop2 = mystack.pop()

numTests += 1
print("Test",numTests,"- two popped values match (30,20)", end="")
if pop1 == 30 and pop2 == 20:
    testPassed += 1
    print("\t\t - PASSED")    

numTests += 1
print("Test",numTests,"- one value left", end="")
if mystack.getCount() == 1:
    testPassed += 1
    print("\t\t\t\t - PASSED")    

top = mystack.top()

numTests += 1
print("Test",numTests,"- last popped value matches", end="")
if top == 10:
    testPassed += 1
    print("\t\t - PASSED")    

pop3 = mystack.pop()

numTests += 1
print("Test",numTests,"- empty after values popped", end="")
if mystack.isEmpty():
    testPassed += 1
    print("\t\t - PASSED")    
    
for i in range(mystack.getSize()):
    mystack.push(i)

numTests += 1
print("Test",numTests,"- overfills", end="")
try: 
    mystack.push(100)
except PracExamException:
    testPassed += 1
    print("\t\t\t\t - PASSED")    
    
        
print("\nNumber of Tests =     ", numTests)
print("Number Tests Passed = ", testPassed)
print("% Passed = ", testPassed*100/numTests)
print(“\n**** Tests Complete ****\n”)
